doc/arch/008-materialized-views-for-dashboards.md

```diff
@@ -0,0 +1,39 @@
+# Materialized views for Dashboards
+
+## Context
+In Dec 2019, we decided to add the new "My Facilities Dashboards" for CVHOs and other health officials.
+These dashboards make a lot more data available to users than the previously existing Cohort reports.
+At the same time, they need to be significantly more performant than the existing reports.
+With the expectation that the size of data would grow rapidly, the following approaches were considered:
+* Extract-Transform-Load data pipeline
+* Materialized Views
+
+## Decision
+An ETL pipeline pushing data to a reporting database is the ideal solution for this purpose but would involve a lot of effort to set up and maintain.
+
+Materialized views are a good intermediate solution, since they give us many of the benefits of an ETL pipeline,
+without the overhead of specialised infrastructure. We may need to review this decision if materialized view refreshes become very expensive.
+We will use [scenic](https://github.com/scenic-views/scenic) to manage the views.
+
+## Approach
+### Defining the views
+* For each dashboard, we may define a view to represent its data, or choose to query existing views.
+* The views will slice data by the time periods we need to query by (eg: `days`, `months`, `quarters`).
+* Ideally, the view definitions should be free of domain logic.
+### Using the views
+* Views will have to be refreshed periodically. The period can vary depending on the dashboard's requirement.
+* Each dashboard will consist of a query object, the necessary materialized views for its data, and templates to render the data appropriately.
+* The query object will provide an interface to fetch the data from the views. These objects should return active record relations to allow for composability and reuse.
+ 
+## Status
+
+Accepted
+
+## Consequences
+
+* Materialized view refresh times will scale with the size of the database tables. Currently `blood_pressures` is the most significant table.
+* View refresh order is important, dependant views should be refreshed only after their parent views to ensure consistency.
```

Comment:
Can we link to the rake/clock task that was created to ensure the views are refreshed in the correct order?

Reply:
PS, have we kept that task up-to-date as we added new views?

Reply:
Added the link, and yes, the task is up-to-date.